src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+        /// </summary>
+        private const string CS0266 = nameof(CS0266);
+
+        /// <summary>
+        /// CS1503: Argument 1: cannot convert from 'x' to 'y'
+        /// </summary>
+        private const string CS1503 = nameof(CS1503);
+
+        /// <summary>
+        /// Give a set of least specific types with a limit, and the part exceeding the limit doesn't show any code fix, but logs telemetry 
+        /// </summary>
+        private const int MaximumConversionOptions = 3;
+
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CS0266, CS1503);
+
+        internal override CodeFixCategory CodeFixCategory => CodeFixCategory.Compile;
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var diagnostic = context.Diagnostics.First();
+
+            var root = await document.GetRequiredSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var targetNode = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true)
+                .GetAncestorsOrThis<ExpressionSyntax>().FirstOrDefault();
+            if (targetNode != null)
+            {
+                var hasSolution = TryGetTargetTypeInfo(semanticModel, diagnostic.Id, targetNode, cancellationToken, out var nodeType, out var potentialConversionTypes);
+                if (!hasSolution)
+                {
+                    return;
+                }
+
+                if (potentialConversionTypes.Length == 1)
+                {
+                    context.RegisterCodeFix(new MyCodeAction(
+                        CSharpFeaturesResources.Add_explicit_cast,
+                        c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+                        context.Diagnostics);
+                }
+                else
+                {
+                    var actions = ArrayBuilder<CodeAction>.GetInstance();
+
+                    // MaximumConversionOptions: we show at most [MaximumConversionOptions] options for this code fixer
+                    for (var i = 0; i < Math.Min(MaximumConversionOptions, potentialConversionTypes.Length); i++)
+                    {
+                        var convType = potentialConversionTypes[i];
+                        actions.Add(new MyCodeAction(string.Format(CSharpFeaturesResources.Convert_type_to_0, convType.ToMinimalDisplayString(semanticModel, context.Span.Start)),
+                            c => Task.FromResult(document.WithSyntaxRoot(ApplyFix(root, targetNode, convType)))));
```

Comment:
```suggestion
                        actions.Add(new MyCodeAction(
                            string.Format(CSharpFeaturesResources.Convert_type_to_0, convType.ToMinimalDisplayString(semanticModel, context.Span.Start)),
                            c => Task.FromResult(document.WithSyntaxRoot(ApplyFix(root, targetNode, convType)))));
```

Reply:
Also, in general, we try not to do anything more than a basic method call after the `c =>` in `new MyCodeAction(`.  Can you extract the body of this to a helper method that this calls.  i.e. `_ => ApplySingleConversionToDocument(...)` or something to that effect?

Reply:
Replace the code by `_ => ApplySingleConversionToDocumentAsync(document, ApplyFix(root, targetNode, convType))`, is this good?